src/mongoose_acc.erl

```diff
@@ -1,300 +1,263 @@
-
-%%%%% internal %%%%%
-
-append(Val, L) when is_list(L), is_list(Val) ->
-    L ++ Val;
-append(Val, L) when is_list(L) ->
-    [Val | L].
-
-dump(_, []) ->
-    ok;
-dump(Acc, [K|Tail]) ->
-    ?ERROR_MSG("~p = ~p", [K, maps:get(K, Acc)]),
-    dump(Acc, Tail).
-
-
-%% @doc pattern-match to figure out (a) which attrs can be 'required' (b) how to cook them
-produce(iq_query_info, Acc) ->
-    Iq = jlib:iq_query_info(get(element, Acc)), % it doesn't change
-    put(iq_query_info, Iq, Acc);
-produce(xmlns, Acc) ->
-    read_children(Acc);
-produce(command, Acc) ->
-    read_children(Acc).
-
-
-%% @doc scan xml children to look for namespace; the name of xml element containing namespace
-%% defines the purpose of a stanza, we store it as 'command'; if absent, we store 'undefined'.
-read_children(Acc) ->
-    Acc1 = put(command, undefined, put(xmlns, undefined, Acc)),
-    #xmlel{children = Children} = get(element, Acc1),
-    read_children(Acc, Children).
-
-read_children(Acc, []) ->
-    Acc;
-read_children(Acc, [#xmlel{} = Chld|Tail]) ->
-    case exml_query:attr(Chld, <<"xmlns">>, undefined) of
-        undefined ->
-            read_children(Acc, Tail);
-        Ns ->
-            #xmlel{name = Name} = Chld,
-            put(command, Name, put(xmlns, Ns, Acc))
-    end;
-read_children(Acc, [_|Tail]) ->
-    read_children(Acc, Tail).
+    Acc#{ Key => V, non_strippable => NewNonStrippable }.
+
+-spec append(NS :: any(), Key :: any(), Val :: any() | [any()], Acc :: t()) -> t().
+append(NS, Key, Val, Acc) ->
+    append(NS, Key, Val, undefined, Acc).
+
+-spec append(NS :: any(),
+            Key :: any(),
+            Val :: any() | [any()],
+            IsStrippable :: strippable(),
+            Acc :: t()) -> t().
+append(NS, Key, Val, IsStrippable, Acc) ->
+    OldVal = get(NS, Key, [], Acc),
+    set(NS, Key, append(OldVal, Val), IsStrippable, Acc).
+
+get(NS, K, #{ mongoose_acc := true } = Acc) ->
+    maps:get({NS, K}, Acc).
+
+get(NS, K, Default, #{ mongoose_acc := true } = Acc) ->
+    maps:get({NS, K}, Acc, Default).
+
+delete(NS, K, #{ mongoose_acc := true, non_strippable := NonStrippable } = Acc0) ->
+    Key = {NS, K},
+    Acc1 = maps:remove(Key, Acc0),
+    Acc1#{ non_strippable => sets:del_element(Key, NonStrippable) }.
+
+%% Doesn't use 'location' param
+-spec strip(ParamsToOverwrite :: strip_params(), Acc :: t()) -> t().
+strip(#{ lserver := NewLServer } = Params,
+      #{ mongoose_acc := true, non_strippable := NonStrippable } = Acc) ->
+    NonStrippableL = sets:to_list(NonStrippable),
+    StrippedAcc = maps:with(NonStrippableL ++ default_non_strippable(), Acc),
+    StrippedAcc#{ lserver => NewLServer, stanza => stanza_from_params(Params) }.
+
+%% --------------------------------------------------------
+%% Debug API
+%% --------------------------------------------------------
+
+-spec dump(t()) -> ok.
+dump(Acc) ->
+    lists:foreach(fun(K) ->
+                          ?ERROR_MSG("~p = ~p", [K, maps:get(K, Acc)])
```

Comment:
Two things here:
1. I'm not sure if error level is good here, maybe warning would be enough?
2. This output can be interleaved with other logs.

I understand this is for debugging purposes so I'm not insisting on changing this.

Reply:
Yes, I'm for removing it.